Add entitlements model to GetFeatures

michellescripts opened the pull request:
### What

![image](https://github.com/gravitational/teleport.e/assets/11967646/7b71ab53-4683-4305-9b68-e071e14adc18)

This is the first set of PRs in a series of PRs to surface [entitlements](https://www.notion.so/goteleport/Entitlement-System-73406dd32fa543ab9a4d67f19a1805d3?pm=c) in Teleport. Entitlements define a customers access to a particular feature, and the source of truth for entitlements is Cloud (sales center / license).

In this PR set we're updating the `GetCloudFeatures` method (previously `FetchFromCloud`) and the `getSelfHostedLicenseFeatures` method (previously `getLicenseFeatures`) to set a new field called `Entitlements.` Entitlements are a map of features to information:

```
AccessRequests: {Enabled: true, Limited: false, Limit: 0}
```

The keys for entitlements are located in `teleport/entitlements.go`

This PR does not propagate any changes past **ClusterFeatures**; future PRs will make similar changes to auth & web config features (reference above image).

**Note** that Identity business logic for enabling and limiting features was added in [Cloud](https://github.com/gravitational/cloud/pull/8783), and features read from cloud should be trusted wrt Identity groupings.

### License

This PR updates the license spec to add a new Entitlements section. Once released, we will be able to update the Cloud repo to start issuing licenses with the new entitlements model.

In [Teleport.E](https://github.com/gravitational/teleport.e/pull/4317
), the `getSelfHostedLicenseFeatures` method is backwards compatible, all the original logic persists and is applied. Entitlements will only be set if they are present on the license (i.e. the instance has a modern license).

### Related

Supports https://github.com/gravitational/cloud/issues/9022
Cloud PR: https://github.com/gravitational/cloud/pull/8783
E PR: https://github.com/gravitational/teleport.e/pull/4317

michellescripts commented on api/types/license.go:
```diff
@@ -156,8 +156,23 @@ type License interface {
+type EntitlementInfo struct {
+	// Enabled indicates the feature is 'on' if true
+	Enabled Bool
+	// Limited indicates the feature is in a 'limited' or 'teaser' state if true
```
Yep, synonym. I'll remove teaser reference in case that was confusing.

michellescripts commented on entitlements.go:
```diff
@@ -0,0 +1,31 @@
+package teleport
+
+// EntitlementKind should be 1:1 with the Features & FeatureStrings in salescenter/product/product.go
+// Custom Theme is dropped; if a string value is set on the parent struct we honor that as the theme
+type EntitlementKind string
```
I have updated the comment, can you re-read and LMK if that makes more sense?

michellescripts commented on api/types/license.go:
```diff
@@ -601,4 +626,7 @@ type LicenseSpecV3 struct {
 	// SupportsPolicy turns Teleport Policy features on or off.
 	SupportsPolicy Bool `json:"policy,omitempty"`
+
+	// entitlements define a customer’s access to a specific features
+	Entitlements map[string]EntitlementInfo `json:"entitlements,omitempty"`
```
I believe that API shouldn't rely on types outside of API. The reading of the license is able to use the EntitlementKind strings to parse the license info, but the license itself is a string key.

kimlisa reviewed (approved):
LGTM

kimlisa commented:
i just had one more clarifying question. how do we determine if flags should be defined outside of entitlement? 

for example `RecoveryCodes` is outside, but flags like `Identity` and `Policy` is part of entitlement map (which i initially thought made more sense outside of it since limits do not apply and these just describe like a product with different features tied to it)

kimlisa commented on lib/auth/auth.go:
```diff
@@ -6854,18 +6855,21 @@ func (a *Server) getAccessRequestMonthlyUsage(ctx context.Context) (int, error)
-		return nil // unlimited
+	accessRequestsEntitlement := f.GetEntitlement(entitlements.AccessRequests)
+
+	if accessRequestsEntitlement.Limit == 0 {
```
i just want us to give this a second thought: https://github.com/gravitational/teleport.e/pull/4317#discussion_r1643912368

michellescripts commented:
> i just had one more clarifying question. how do we determine if flags should be defined outside of entitlement? for example RecoveryCodes is outside, but flags like Identity and Policy is part of entitlement map (which i initially thought made more sense outside of it since limits do not apply and these just describe like a product with different features tied to it)

@kimlisa I've added some [comments](https://github.com/gravitational/teleport/pull/42394/files#diff-8e9da4d6acbdd048ce7ceafde7ee7934d51890dc15475a71644a7a76f901884dR65) to come back and re-evaluate that logic. I don't want to make too many breaking changes here but I do believe it is very confusing to mix entitlements (access to feature) with non entitlement flags. It could be that some of these end up being true entitlements which are set by Cloud or they could be deprecated and or replaced with other logic.

Long term the goal is to remove "Feature Grouping" concepts like `Identity` from the Teleport side and to limit the amount of grouping we support in Sales Center. The [identity](https://github.com/gravitational/cloud/pull/8783/files#diff-0053cd0a4494ce96365072b8d4dfcf2c7b74e212ebe9a2fba19395bc49b1eeecR111) grouping is already being deconstructed in entitlements and hopefully we can start to rely on the entitlement itself where applicable.

kimlisa commented on lib/modules/modules.go:
```diff
@@ -38,184 +38,130 @@ import (
+	// AccessMonitoringFeature enables the usage of access monitoring.
+	// NOTE: this flag is used to signal that Access Monitoring is *enabled* on a cluster.
+	// *Access* to the feature is gated on the `AccessMonitoring` entitlement.
+	AccessMonitoringFeature bool
```
should we just rename this to `AccessMonitoringConfigured`? b/c i think `AccessMonitoringFeature` sounds too much like `is cluster entitled to this feature` (up to you)

kimlisa commented on api/types/license.go:
```diff
@@ -601,4 +624,7 @@ type LicenseSpecV3 struct {
 	// SupportsPolicy turns Teleport Policy features on or off.
 	SupportsPolicy Bool `json:"policy,omitempty"`
+
+	// entitlements define a customer’s access to a specific features
```
should we add a comment somewhere that explains how feature support checking should be done?

eg:
```go
{ AccessMonitoring: true, Limit: 0} = full support
{ AccessMonitoring: true, Limit: LIMIT } = limited support
{ AccessMonitoring: false, limit: LIMIT } = no support 
```

to check for `full support` you'll need to check for `Entitled == true + 0 limit`

michellescripts replied:
This is documented in a sense on line 168 where we define the EntitlementInfo struct:

```
// EntitlementInfo is the state and limits of a particular entitlement
type EntitlementInfo struct {
	// Enabled indicates the feature is 'on' if true; feature is disabled if false
	Enabled Bool
	// Limit indicates the allotted amount of use when limited; if 0 use is unlimited
	Limit int32
}
```

Does this suffice or do you want me to add the explicit examples?

```
// EntitlementInfo is the state and limits of a particular entitlement; Examples:
// { FeatureX: true, Limit: 0} = unlimited access to FeatureX
// { FeatureX: true, Limit: LIMIT } = limited access to FeatureX
// { FeatureX: false, limit: LIMIT } = no access to FeatureX
type EntitlementInfo struct {
	// Enabled indicates the feature is 'on' if true; feature is disabled if false
	Enabled Bool
	// Limit indicates the allotted amount of use when limited; if 0 use is unlimited
	Limit int32
}
```

kimlisa replied:
i appreciate the explicit 👍

michellescripts replied:
Added examples in https://github.com/gravitational/teleport/pull/42394/commits/e7762b78fd40fbf99338f7311fb201b7b4fa040e

michellescripts committed:
Surface Cloud Entitlements

- Cloud entitlements are now available on Features
- Legacy logic applies if entitlements are not present

zmb3 commented:
/excludeflake *
